src/main/java/sdmExam/Board.java

```diff
@@ -80,4 +87,48 @@ protected Stone colorWithCompleteChain() {
     protected Stream<Intersection> getEmptyIntersections() {
         return intersections.stream().filter(intersection -> !intersection.isOccupied());
     }
+
+    //TODO: don't know if it's useful to get the territories or just act on them
+    public List<Set<Intersection>> getTerritories() {
+        return regionsContainer.getRegions().stream()
+                .filter(region -> region.stream().allMatch(this::isOrthogonalToAtLeastTwoStones))
+                .collect(Collectors.toList());
+    }
+
+    private boolean isOrthogonalToAtLeastTwoStones(Intersection intersection) {
+        return intersections.stream()
+                .filter(intersection::isOrthogonalTo)
+                .filter(Intersection::isOccupied)
+                .count() >= 2;
+    }
+
+    public List<Intersection> getOrthogonalAdjacencyIntersections(Intersection intersection) {
+        return intersections.stream()
+                .filter(otherIntersection -> otherIntersection.isOrthogonalTo(intersection))
+                .collect(Collectors.toList());
+    }
+
+    public void fillTerritory(List<Intersection> territory, Player lastPlay){
```

Comment:
Still a little bit long, I will take a look at it in the afternoon. The name's a bit misleading: the actual filling operation is done at line 132, the rest is understand which is the correct color to fill.

Reply:
I just tried to refactor a bit more. Please consider the latest version